File: legacykit-ui/src/lib/api/shsh.ts
import { invoke } from '@tauri-apps/api/core';

export interface SaveShshRequest {
  deviceType: string;
  deviceEcid: string;
  boardConfig: string | null;
  iosVersion: string;
  buildId: string | null;
  buildManifestPath: string | null;
  apnonce: string | null;
  generator: string | null;
  outputDir: string;
}

export interface SaveShshResult {
  blobPaths: string[];
  args: string[];
}

export interface CydiaBlobRequest {
  deviceType: string;
  deviceEcid: string;
  buildIds: string[];
  outputDir: string;
}

export interface CydiaBlobAttempt {
  buildId: string;
  saved: boolean;
  blobPath: string | null;
  message: string | null;
}

export interface CydiaBlobResult {
  attempts: CydiaBlobAttempt[];
}

export interface DumpOnboardBlobRequest {
  rawDumpPath: string;
  outputPath: string;
}

export interface DumpOnboardBlobResult {
  blobPath: string;
  args: string[];
}

export interface ListBlobsRequest {
  directory: string;
}

export interface SavedBlob {
  path: string;
  fileName: string;
  sizeBytes: number;
  modifiedUnix: number | null;
  deviceType: string | null;
  deviceEcid: string | null;
  iosVersion: string | null;
  buildId: string | null;
}

export interface ListBlobsResult {
  directory: string;
  blobs: SavedBlob[];
}

export function saveShshBlob(request: SaveShshRequest): Promise<SaveShshResult> {
  return invoke<SaveShshResult>('save_shsh_blob', { request });
}

export function fetchCydiaBlobs(request: CydiaBlobRequest): Promise<CydiaBlobResult> {
  return invoke<CydiaBlobResult>('fetch_cydia_blobs', { request });
}

export function dumpOnboardBlob(
  request: DumpOnboardBlobRequest,
): Promise<DumpOnboardBlobResult> {
  return invoke<DumpOnboardBlobResult>('dump_onboard_blob', { request });
}

export function listSavedBlobs(request: ListBlobsRequest): Promise<ListBlobsResult> {
  return invoke<ListBlobsResult>('list_saved_blobs', { request });
}

File: legacykit-ui/src-tauri/src/models/shsh.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveShshRequest {
    pub device_type: String,
    pub device_ecid: String,
    pub board_config: Option<String>,
    pub ios_version: String,
    pub build_id: Option<String>,
    pub build_manifest_path: Option<String>,
    pub apnonce: Option<String>,
    pub generator: Option<String>,
    pub output_dir: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveShshResult {
    pub blob_paths: Vec<String>,
    pub args: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CydiaBlobRequest {
    pub device_type: String,
    pub device_ecid: String,
    pub build_ids: Vec<String>,
    pub output_dir: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CydiaBlobAttempt {
    pub build_id: String,
    pub saved: bool,
    pub blob_path: Option<String>,
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CydiaBlobResult {
    pub attempts: Vec<CydiaBlobAttempt>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DumpOnboardBlobRequest {
    pub raw_dump_path: String,
    pub output_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DumpOnboardBlobResult {
    pub blob_path: String,
    pub args: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListBlobsRequest {
    pub directory: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SavedBlob {
    pub path: String,
    pub file_name: String,
    pub size_bytes: u64,
    pub modified_unix: Option<u64>,
    pub device_type: Option<String>,
    pub device_ecid: Option<String>,
    pub ios_version: Option<String>,
    pub build_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListBlobsResult {
    pub directory: String,
    pub blobs: Vec<SavedBlob>,
}
